lib/util/git/auth.ts

```diff
@@ -115,33 +116,45 @@ function getAuthenticationRulesWithToken(
  */
 export function getAuthenticationRules(
   gitUrl: string,
+  hostType: string | undefined | null,
   token: string,
 ): AuthenticationRule[] {
   const authenticationRules = [];
   const hasUser = token.split(':').length > 1;
   const insteadUrl = parseGitUrl(gitUrl);
 
+  if (hostType === 'bitbucket-server') {
+    insteadUrl.source = 'bitbucket-server';
+  }
+
   const url = { ...insteadUrl };
   const protocol = regEx(/^https?$/).test(url.protocol)
     ? url.protocol
     : 'https';
 
+  const sshUrl = { ...insteadUrl };
```

Comment:
What's the need/benefit of creating `sshUrl` instead of reusing `insteadUrl`?

Reply:
Because `insteadUrl` is used (by the existing code) later to generate a `https://` url

Reply:
I think this function has now crossed the line of being too complex, due to `insteadUrl`, `url` and `sshUrl` all being similar but the relationships undocumented. Do you think a refactor can be done to reduce or simplify this, or we just need to document it?

Reply:
What if I rewrite it like this? https://github.com/renovatebot/renovate/pull/29951/commits/cd401892d17f9ef71b6f72977abb90e9a6f47b46

Reply:
Yes, that's better. But please take the opportunity to document this logic so it's easier to maintain in future (including any confusing parts which were there already)

Reply:
I'm unable to write any useful documentation because I don't fully understand this code. For example, is this function expected to take only https urls, or ssh urls too. On one hand, it seems to accept non-http on line 144. On the other hand the comment on line 166 suggests `protocol` is expected to be `https` and nothing else. Also from the surrounding code and tests it looks like only https is expected.

Plus half of the logic (including special cases/workarounds) is in git-url-parse, and half is in Renovate. For example, see [`buildToken`](https://github.com/IonicaBizau/git-url-parse/blob/28828546c148d58bbcff61409915a4e1e8f7eb11/lib/index.js#L293) in git-url-parse and `getAuthenticationRulesWithToken` here.